src/transport/iroh_net.rs

```diff
@@ -0,0 +1,750 @@
+        loop {
+            tracing::debug!("Waiting for incoming connection...");
+            let connecting = match endpoint.accept().await {
+                Some(connecting) => connecting,
+                None => break,
+            };
+
+            tracing::debug!("Awaiting connection from connect...");
+            let connection = match connecting.await {
+                Ok(connection) => connection,
+                Err(e) => {
+                    tracing::warn!(?e, "Error accepting connection");
+                    continue;
+                }
+            };
+
+            // When the `allowed_node_ids` is empty, it's empty forever, so the CPU's branch
+            // prediction should always optimize this block away from this loop.
+            // The same applies when it isn't empty, ignoring the check for emptiness and always
+            // extracting the node id and checking if it's in the set.
+            if !allowed_node_ids.is_empty() {
+                let Ok(client_node_id) = iroh_net::endpoint::get_remote_node_id(&connection)
+                    .map_err(|e| {
+                        tracing::error!(
+                            ?e,
+                            "Failed to extract iroh-net node id from incoming connection from {:?}",
+                            connection.remote_address()
+                        )
+                    })
+                else {
+                    connection.close(0u32.into(), b"failed to extract iroh-net node id");
+                    continue;
+                };
+
+                if !allowed_node_ids.contains(&client_node_id) {
+                    connection.close(0u32.into(), b"forbidden node id");
+                    continue;
+                }
+            }
+
+            tracing::debug!(
+                "Connection established from {:?}",
+                connection.remote_address()
+            );
+
+            tracing::debug!("Spawning connection handler...");
+            tokio::spawn(Self::connection_handler(connection, sender.clone()));
+        }
+    }
+
+    /// Create a new server channel, given a quinn endpoint, with unrestricted access by node id
+    ///
+    /// The server channel will take care of listening on the endpoint and spawning
+    /// handlers for new connections.
+    pub fn new(endpoint: iroh_net::Endpoint) -> io::Result<Self> {
+        Self::new_with_access_control(endpoint, AccessControl::Unrestricted)
+    }
+
+    /// Create a new server endpoint, with specified access control
+    ///
+    /// The server channel will take care of listening on the endpoint and spawning
+    /// handlers for new connections.
+    pub fn new_with_access_control(
+        endpoint: iroh_net::Endpoint,
+        access_control: AccessControl,
+    ) -> io::Result<Self> {
+        let allowed_node_ids = match access_control {
+            AccessControl::Unrestricted => BTreeSet::new(),
+            AccessControl::Allowed(list) if list.is_empty() => {
```

Comment:
Why have this check at all? Having an empty allow list just means that you don't allow any incoming connections. That's a perfectly valid thing to do, even if it is a bit weird.

Reply:
Because I think that it's better to fail early at setup time than have to debug out the reason of why the server is rejecting all requests. Having to debug it out until you find out that you forgot to load an env var with a list of node ids for example, having to do a huge facepalm when you figure it out haha. 
Also, due to the way that I've implemented the rejection part, if the `BTreeSet` is empty there, I treat it as unrestricted access, not as a reject all. Implementing it like this brings 2 possible benefits, the first one being that when you want unrestricted access, due to the CPU's branch predictor, that entire block of code will be skipped out. And the second is basically what I mentioned above, having a server being unresponsive can be a hard to debug issue, aside from being a useless server lol

Reply:
Hm, not entirely convinced. Imagine you got a node that supports rpc via iroh connections, but you want to disable it most of the time and only enable it on demand. Isn't using an empty list of allowed nodes a very reasonable thing to do in this case?

If I want to disallow ssh access, I have an empty `authorized_keys` file or none at all. This is not a misconfiguration, but fully intentional.

Reply:
But the current implementation doesn't allow for dynamically adding or removing authorized nodes, once you instantiate the server, it will use those authorized node ids until you drop the server. I can implement like this if it's considered an important feature. Although like on the server away and back test, you can recreate the server as many times as you want, attempting to connect to a non-existing server will have the same effect than attempting to connect to a server which will refuse all connections.